### PR TITLE
Add training code for running on colab

### PR DESCRIPTION
**Pull request by m41w4r3exe**
- Add trainer.py for training on colab
- Add method comments for decoder

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -4,7 +4,14 @@
   // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
   "version": "0.2.0",
   "configurations": [
-
+    {
```
Comment (louis-demetz) [inline .vscode/launch.json:7]: Does this file really belong on github? isn't this rather a personal ide configuration file?

File: source/decoder.py [hunk 2]
```diff
@@ -1,16 +1,37 @@
...
+    1- List of events
+    2- Then converts these events to midi file via MidiTok and miditoolkit
+
+    :param tokenizer: from MidiTok
+
+    Usage with write_to_midi method:
+        args: text(String) example ->  PIECE_START TRACK_START INST=25 DENSITY=2 BAR_START NOTE_ON=50 TIME_DELTA=1 NOTE_OFF=50...BAR_END TRACK_END
+        returns: midi file from miditoolkit
+    """
+
```
Comment (louis-demetz) [inline source/decoder.py:23]: Fancy comments are nice <3

File: source/trainer.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
+# RUN 3 lines below in a seperate cell in Google Colab
+# !pip install transformers tokenizers wandb huggingface_hub
+# from huggingface_hub import notebook_login
+# notebook_login()
+
```
Comment (louis-demetz) [inline source/trainer.py:5]: If you wanna run this on colab, shouldn't this file be a .ipynb ?

Comment (misnaej) [reply source/trainer.py:5]: We will import the scripts from the cloned repo directly in collab

File: source/trainer.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+from tokenizers.models import WordLevel
+from tokenizers.pre_tokenizers import WhitespaceSplit
+from tokenizers.trainers import WordLevelTrainer
+from datetime import datetime
+import wandb
+
+# CONFIG:
+TRAIN_FROM_CHECKPOINT = None  # Example: checkpoint-80000
+EVAL_STEPS = 1000
+PER_DEVICE_TRAIN_BATCH_SIZE = 10
```
Comment (misnaej) [inline source/trainer.py:26]: the default value should be much lower. 10 sometimes broke collab even with a small dataset
PER_DEVICE_TRAIN_BATCH_SIZE = 2 or 4

File: source/trainer.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    print("----")
+    print(data)
+    print(dataset_train_tokenized[i])
+
+model_config = GPT2Config(
+    vocab_size=tokenizer.vocab_size,
+    pad_token_id=tokenizer.pad_token_id,
+    n_embd=512,
+    n_head=8,
+    n_layer=10,
```
Comment (misnaej) [inline source/trainer.py:104]: maybe n_layer could be smaller, like n_layer = 8 ; that will speed up the training. we should check whether it makes a big difference

File: source/trainer.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+from tokenizers import Tokenizer
+from tokenizers.models import WordLevel
+from tokenizers.pre_tokenizers import WhitespaceSplit
+from tokenizers.trainers import WordLevelTrainer
+from datetime import datetime
+import wandb
+
+# CONFIG:
+TRAIN_FROM_CHECKPOINT = None  # Example: checkpoint-80000
+EVAL_STEPS = 1000
```
Comment (misnaej) [inline source/trainer.py:25]: if we have >300_000 steps I would suggest to set this much higher... 
maybe EVAL_STEPS could be by default at least 2 % of the total amount of steps?

File: source/trainer.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+if TRAIN_FROM_CHECKPOINT is not None:
+    trainer.train(f"{model_path}/{TRAIN_FROM_CHECKPOINT}")
+else:
+    trainer.train()
+
+
+# Save the tokenizer.
+tokenizer.save_pretrained(model_path)
+model.save_pretrained(model_path)
+trainer.push_to_hub()
```
Comment (misnaej) [inline source/trainer.py:147]: Nice!
I guess we will put everything into Classes and function once it runs well

File: source/trainer.py (hunk 3 above)
Comment (louis-demetz) [reply source/trainer.py:5]: ok makes sense

**Review by misnaej — APPROVED** [review]

**Pull request merged**